{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext watermark\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools as it\n",
    "\n",
    "from Bio import Phylo as BioPhylo\n",
    "import alifedata_phyloinformatics_convert as apc\n",
    "import covasim as cv\n",
    "from hstrat import _auxiliary_lib as hstrat_aux\n",
    "from IPython.core.display import display, HTML\n",
    "from matplotlib import patches as mpl_patches\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats as scipy_stats\n",
    "import seaborn as sns\n",
    "from slugify import slugify\n",
    "from teeplot import teeplot as tp\n",
    "\n",
    "from pylib._cv_infection_log_to_alstd_df import cv_infection_log_to_alstd_df\n",
    "from pylib._draw_biophylo import draw_biophylo\n",
    "from pylib._generate_dummy_sequences_phastSim import (\n",
    "    generate_dummy_sequences_phastSim,\n",
    ")\n",
    "from pylib._mask_mutations import mask_mutations\n",
    "from pylib._mwu_kdeplot import mwu_kdeplot\n",
    "from pylib._pairplot import pairplot\n",
    "from pylib._screen_mutation_defined_nodes import screen_mutation_defined_nodes\n",
    "from pylib._seed_global_rngs import seed_global_rngs\n",
    "from pylib._SyncHostCompartments import SyncHostCompartments\n",
    "from pylib._VariantFlavor import VariantFlavor\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%watermark -diwmuv -iv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "teeplot_subdir = \"2025-04-11-dev-sequenceconfig\"\n",
    "teeplot_subdir\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup CovaSim Params\n",
    "\n",
    "Define `gamma` and `delta` strains, as well as accompanying lower-transmissibility `gamma'` and `delta'` strains.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_size = 20_000\n",
    "num_days = 200\n",
    "num_variants = 5  # 0th variant is wild-type\n",
    "seed_global_rngs(1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://github.com/InstituteforDiseaseModeling/covasim/blob/fef033fed5a0ca71be9c4f773d75dce1d237e753/covasim/parameters.py\n",
    "gamma = dict(\n",
    "    rel_beta=2.05,  # Estimated to be 1.7-2.4-fold more transmissible than wild-type: https://science.sciencemag.org/content/early/2021/04/13/science.abh2644\n",
    "    rel_symp_prob=1.0,\n",
    "    rel_severe_prob=2.6,  # From https://www.eurosurveillance.org/content/10.2807/1560-7917.ES.2021.26.16.2100348\n",
    "    rel_crit_prob=1.0,\n",
    "    rel_death_prob=1.0,\n",
    ")\n",
    "gamma_ = {**gamma, \"rel_beta\": 0.5}  # lower-transmissibility variant\n",
    "\n",
    "delta = dict(\n",
    "    rel_beta=2.2,  # Estimated to be 1.25-1.6-fold more transmissible than B117: https://www.researchsquare.com/article/rs-637724/v1\n",
    "    rel_symp_prob=1.0,\n",
    "    rel_severe_prob=3.2,  # 2x more transmissible than alpha from https://mobile.twitter.com/dgurdasani1/status/1403293582279294983\n",
    "    rel_crit_prob=1.0,\n",
    "    rel_death_prob=1.0,\n",
    ")\n",
    "delta_ = {**delta, \"rel_beta\": 0.6}  # lower-transmissibility variant\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_sequences = pd.read_csv(\"https://osf.io/hp25c/download\")\n",
    "sequence_lookup = dict(\n",
    "    zip(\n",
    "        reference_sequences[\"WHO Label\"].values,\n",
    "        reference_sequences[\"Aligned Sequence\"].values,\n",
    "    ),\n",
    ")\n",
    "variant_flavors = [\n",
    "    VariantFlavor(\n",
    "        variant_wt=cv.variant(\n",
    "            variant=gamma, label=\"Gamma+\", days=0, n_imports=10\n",
    "        ),\n",
    "        variant_mut=cv.variant(\n",
    "            variant=gamma_, label=\"Gamma'\", days=0, n_imports=0\n",
    "        ),\n",
    "        withinhost_r_wt=2.0,\n",
    "        withinhost_r_mut=3.0,\n",
    "        active_strain_factor_wt=2.0,\n",
    "        active_strain_factor_mut=3.0,\n",
    "        p_wt_to_mut=0.01,\n",
    "        sequence=sequence_lookup[\"Gamma\"],\n",
    "    ),\n",
    "    VariantFlavor(\n",
    "        variant_wt=cv.variant(\n",
    "            variant=delta, label=\"Delta+\", days=0, n_imports=10\n",
    "        ),\n",
    "        variant_mut=cv.variant(\n",
    "            variant=delta_, label=\"Delta'\", days=0, n_imports=0\n",
    "        ),\n",
    "        withinhost_r_wt=2.0,\n",
    "        withinhost_r_mut=3.0,\n",
    "        active_strain_factor_wt=2.0,\n",
    "        active_strain_factor_mut=3.0,\n",
    "        p_wt_to_mut=0.01,\n",
    "        sequence=sequence_lookup[\"Delta\"],\n",
    "    ),\n",
    "]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Example Simulation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim = cv.Sim(\n",
    "    interventions=[\n",
    "        SyncHostCompartments(\n",
    "            variant_flavors=variant_flavors,\n",
    "            pop_size=pop_size,\n",
    "        ),\n",
    "    ],\n",
    "    n_days=num_days,\n",
    "    # adapted from https://docs.idmod.org/projects/covasim/en/latest/tutorials/tut_immunity.html\n",
    "    nab_decay=dict(\n",
    "        form=\"nab_growth_decay\",\n",
    "        growth_time=21,\n",
    "        decay_rate1=0.1,\n",
    "        decay_time1=47,\n",
    "        decay_rate2=0.05,\n",
    "        decay_time2=106,\n",
    "    ),\n",
    "    pop_infected=0,  # disable wild-type strain\n",
    "    pop_size=pop_size,\n",
    "    variants=[\n",
    "        v for vf in variant_flavors for v in (vf.variant_wt, vf.variant_mut)\n",
    "    ],\n",
    "    rand_seed=1,\n",
    ")\n",
    "sim.run()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tp.tee(\n",
    "    sim.plot,\n",
    "    \"variant\",\n",
    "    teeplot_outattrs={\"what\": \"within-host\"},\n",
    "    teeplot_subdir=teeplot_subdir,\n",
    ")\n",
    "...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualize Pathogen Phylogeny\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_df = cv_infection_log_to_alstd_df(sim.people.infection_log)\n",
    "\n",
    "phylo_df = hstrat_aux.alifestd_downsample_tips_asexual(\n",
    "    phylo_df, n_downsample=200\n",
    ")\n",
    "\n",
    "phylo_tree = apc.alife_dataframe_to_biopython_tree(\n",
    "    phylo_df,\n",
    "    setattrs=[\"variant\"],\n",
    "    setup_branch_lengths=True,\n",
    ")\n",
    "\n",
    "palette = sns.color_palette(\"Set2\", n_colors=len(variant_flavors) * 2).as_hex()\n",
    "palette_iter = it.cycle(palette)\n",
    "node_colors = {\n",
    "    variant.label: next(palette_iter)\n",
    "    for vf in variant_flavors\n",
    "    for variant in (vf.variant_wt, vf.variant_mut)\n",
    "}\n",
    "for node in phylo_tree.find_clades():\n",
    "    node.color = node_colors[node.variant]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with tp.teed(\n",
    "    draw_biophylo,\n",
    "    phylo_tree,\n",
    "    teeplot_outattrs={\"what\": \"within-host\"},\n",
    "    teeplot_outdir=teeplot_subdir,\n",
    ") as ax:\n",
    "    ax.set_xlabel(\"Days\")\n",
    "    ax.yaxis.set_visible(False)\n",
    "    ax.spines[\"left\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "\n",
    "    legend_handles = [\n",
    "        mpl_patches.Patch(color=color, label=variant)\n",
    "        for variant, color in node_colors.items()\n",
    "    ]\n",
    "    ax.legend(\n",
    "        handles=legend_handles,\n",
    "        loc=\"upper center\",\n",
    "        bbox_to_anchor=(0.5, 1.15),\n",
    "        ncol=4,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Pathogen Phylogeny Topology\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_df = cv_infection_log_to_alstd_df(sim.people.infection_log)\n",
    "phylo_df = hstrat_aux.alifestd_to_working_format(phylo_df, mutate=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean tree topology\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"alifestd_add_inner_leaves\",\n",
    "    logger=print,\n",
    "):\n",
    "    phylo_df = hstrat_aux.alifestd_add_inner_leaves(phylo_df, mutate=True)\n",
    "    phylo_df = hstrat_aux.alifestd_to_working_format(phylo_df, mutate=True)\n",
    "\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"alifestd_downsample_tips_asexual\",\n",
    "    logger=print,\n",
    "):\n",
    "    # phylo_df = hstrat_aux.alifestd_downsample_tips_asexual(\n",
    "    #     phylo_df,\n",
    "    #     n_downsample=2_000,\n",
    "    # )\n",
    "    phylo_df = hstrat_aux.alifestd_to_working_format(phylo_df, mutate=True)\n",
    "\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"alifestd_splay_polytomies\",\n",
    "    logger=print,\n",
    "):\n",
    "    phylo_df = hstrat_aux.alifestd_splay_polytomies(phylo_df, mutate=True)\n",
    "    phylo_df.drop(columns=[\"is_leaf\"], inplace=True, errors=\"ignore\")\n",
    "    phylo_df = hstrat_aux.alifestd_to_working_format(phylo_df, mutate=True)\n",
    "\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"alifestd_collapse_unifurcations\",\n",
    "    logger=print,\n",
    "):\n",
    "    phylo_df = hstrat_aux.alifestd_collapse_unifurcations(\n",
    "        phylo_df, mutate=True\n",
    "    )\n",
    "    phylo_df = hstrat_aux.alifestd_to_working_format(phylo_df, mutate=True)\n",
    "\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"alifestd_delete_unifurcating_roots_asexual\",\n",
    "    logger=print,\n",
    "):\n",
    "    phylo_df = hstrat_aux.alifestd_delete_unifurcating_roots_asexual(\n",
    "        phylo_df, mutate=True\n",
    "    )\n",
    "    phylo_df = hstrat_aux.alifestd_to_working_format(phylo_df, mutate=True)\n",
    "\n",
    "# more statistics\n",
    "with hstrat_aux.log_context_duration(\"alifestd_mark_*\", logger=print):\n",
    "    phylo_df = hstrat_aux.alifestd_mark_leaves(phylo_df, mutate=True)\n",
    "    phylo_df = hstrat_aux.alifestd_mark_num_leaves_sibling_asexual(\n",
    "        phylo_df, mutate=True\n",
    "    )\n",
    "    phylo_df = hstrat_aux.alifestd_mark_num_leaves_asexual(\n",
    "        phylo_df, mutate=True\n",
    "    )\n",
    "    phylo_df = hstrat_aux.alifestd_mark_roots(phylo_df, mutate=True)\n",
    "\n",
    "phylo_df[\"is_leaf\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute Sister Statistics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_leaves = 4\n",
    "work_mask = (phylo_df[\"num_leaves\"] > min_leaves) & (\n",
    "    phylo_df[\"num_leaves_sibling\"] > min_leaves\n",
    ")\n",
    "# sister statistics\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"alifestd_mark_clade_subtended_duration_ratio_sister_asexual\",\n",
    "    logger=print,\n",
    "):\n",
    "    phylo_df = (\n",
    "        hstrat_aux.alifestd_mark_clade_subtended_duration_ratio_sister_asexual(\n",
    "            phylo_df, mutate=True\n",
    "        )\n",
    "    )\n",
    "    phylo_df[\"clade duration ratio\"] = np.log(\n",
    "        phylo_df[\"clade_subtended_duration_ratio_sister\"]\n",
    "    )\n",
    "\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"alifestd_mark_clade_logistic_growth_sister_asexual\",\n",
    "    logger=print,\n",
    "):\n",
    "    phylo_df = hstrat_aux.alifestd_mark_clade_logistic_growth_sister_asexual(\n",
    "        phylo_df, mutate=True, parallel_backend=\"loky\", work_mask=work_mask\n",
    "    )\n",
    "    phylo_df[\"clade growth ratio\"] = phylo_df[\"clade_logistic_growth_sister\"]\n",
    "\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"alifestd_mark_clade_leafcount_ratio_sister_asexual\",\n",
    "    logger=print,\n",
    "):\n",
    "    phylo_df = hstrat_aux.alifestd_mark_clade_leafcount_ratio_sister_asexual(\n",
    "        phylo_df, mutate=True\n",
    "    )\n",
    "    phylo_df[\"clade size ratio\"] = np.log(\n",
    "        phylo_df[\"clade_leafcount_ratio_sister\"]\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate Pathogen Dummy Sequences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_sequences = pd.read_csv(\"https://osf.io/hp25c/download\")\n",
    "ancestral_sequence = reference_sequences.loc[\n",
    "    reference_sequences[\"WHO Label\"] == \"Wildtype\", \"Aligned Sequence\"\n",
    "].squeeze()\n",
    "\n",
    "# strip whtiespace\n",
    "ancestral_sequence = \"\".join(ancestral_sequence.split())\n",
    "\n",
    "# shorten\n",
    "ancestral_sequence = ancestral_sequence[:20]\n",
    "\n",
    "phylo_df[\"variant_flavor\"] = \"wildtype\"\n",
    "dummy_sequences = generate_dummy_sequences_phastSim(\n",
    "    phylo_df,\n",
    "    ancestral_sequences={\"wildtype\": ancestral_sequence},\n",
    ")\n",
    "dummy_sequences[\"variant\"] = dummy_sequences[\"id\"].map(\n",
    "    phylo_df.set_index(\"id\")[\"variant\"].to_dict(),\n",
    ")\n",
    "\n",
    "ancestral_sequence = \"A\" + ancestral_sequence\n",
    "dummy_sequences[\"sequence\"] = (\n",
    "    dummy_sequences[\"variant\"]\n",
    "    .str.contains(\"'\")\n",
    "    .map(\n",
    "        {\n",
    "            True: \"X\",\n",
    "            False: \"A\",\n",
    "        },\n",
    "    )\n",
    "    + dummy_sequences[\"sequence\"]\n",
    ")\n",
    "\n",
    "display(dummy_sequences.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Mutation Masks, Identify Mutaiton-Defined Clades\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_masks = mask_mutations(\n",
    "    ancestral_sequence=ancestral_sequence,\n",
    "    derived_sequences=dummy_sequences[\"sequence\"],\n",
    ")\n",
    "\n",
    "# convert leaf masks to phylo masks\n",
    "for k, v in mutation_masks.items():  # leaf masks\n",
    "    phylo_mask = np.zeros(len(phylo_df), dtype=bool)\n",
    "    phylo_mask[dummy_sequences.loc[v, \"id\"].values] = True\n",
    "    mutation_masks[k] = phylo_mask\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process Pathogen Phylogeny Statistics --- TFP\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation = (0, \"A\", \"X\")\n",
    "display(HTML(f\"<h2>{mutation=}</h2>\"))\n",
    "\n",
    "# screen phylogeny\n",
    "with hstrat_aux.log_context_duration(\n",
    "    \"screen_mutation_defined_nodes\",\n",
    "    logger=print,\n",
    "):\n",
    "    screen = screen_mutation_defined_nodes(\n",
    "        phylo_df.copy(),\n",
    "        has_mutation=mutation_masks[mutation],\n",
    "    )\n",
    "\n",
    "print(screen[\"fisher\"].sum(), screen[\"naive\"].sum(), screen[\"combined\"].sum())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# draw screened phylogeny, color by clades\n",
    "draw_df = phylo_df.copy()\n",
    "draw_df[\"is_defining\"] = screen[\"combined\"]\n",
    "draw_df = hstrat_aux.alifestd_mask_descendants_asexual(\n",
    "    draw_df, ancestor_mask=screen[\"combined\"], mutate=True\n",
    ")\n",
    "draw_df[\"is_screened\"] = draw_df[\"alifestd_mask_descendants_asexual\"]\n",
    "draw_df[\"extant\"] = draw_df[\"is_screened\"]\n",
    "draw_df = hstrat_aux.alifestd_prune_extinct_lineages_asexual(draw_df)\n",
    "draw_df = hstrat_aux.alifestd_to_working_format(draw_df)\n",
    "draw_df = hstrat_aux.alifestd_collapse_unifurcations(draw_df)\n",
    "draw_df = hstrat_aux.alifestd_to_working_format(draw_df)\n",
    "print(f\"{len(draw_df)=}\")\n",
    "draw_tree = apc.alife_dataframe_to_biopython_tree(\n",
    "    draw_df,\n",
    "    setattrs=[\"variant\", \"is_defining\"],\n",
    "    setup_branch_lengths=True,\n",
    ")\n",
    "for node in draw_tree.find_clades():\n",
    "    node.color = [\n",
    "        node_colors[node.variant],\n",
    "        node_colors[node.variant],\n",
    "        # \"magenta\",\n",
    "    ][node.is_defining]\n",
    "\n",
    "with tp.teed(\n",
    "    draw_biophylo,\n",
    "    draw_tree,\n",
    "    label_func=lambda node: [None, \"X\"][node.is_defining],\n",
    "    teeplot_outattrs={\n",
    "        \"color\": \"defining-mut-clades\",\n",
    "        \"what\": slugify(str(mutation)),\n",
    "    },\n",
    "    teeplot_outdir=teeplot_subdir,\n",
    ") as ax:\n",
    "    ax.set_xlabel(\"Days\")\n",
    "    ax.yaxis.set_visible(False)\n",
    "    ax.spines[\"left\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "\n",
    "    legend_handles = [\n",
    "        mpl_patches.Patch(color=color, label=variant)\n",
    "        for variant, color in node_colors.items()\n",
    "    ]\n",
    "    ax.legend(\n",
    "        handles=legend_handles,\n",
    "        loc=\"upper center\",\n",
    "        bbox_to_anchor=(0.5, 1.15),\n",
    "        ncol=4,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# draw screened phylogeny, color byscreened status\n",
    "draw_df = phylo_df.copy()\n",
    "draw_df = hstrat_aux.alifestd_mask_descendants_asexual(\n",
    "    draw_df,\n",
    "    ancestor_mask=screen[\"combined\"],\n",
    "    mutate=True,\n",
    ")\n",
    "draw_df[\"is_screened\"] = draw_df[\"alifestd_mask_descendants_asexual\"]\n",
    "draw_df = hstrat_aux.alifestd_mask_descendants_asexual(\n",
    "    draw_df,\n",
    "    ancestor_mask=screen[\"combined\"][draw_df[\"sister_id\"]],\n",
    "    mutate=True,\n",
    ")\n",
    "draw_df[\"sister_screened\"] = draw_df[\"alifestd_mask_descendants_asexual\"]\n",
    "draw_df[\"extant\"] = draw_df[\"is_screened\"] | draw_df[\"sister_screened\"]\n",
    "draw_df = hstrat_aux.alifestd_prune_extinct_lineages_asexual(draw_df)\n",
    "draw_df = hstrat_aux.alifestd_to_working_format(draw_df)\n",
    "draw_df = hstrat_aux.alifestd_collapse_unifurcations(draw_df)\n",
    "draw_df = hstrat_aux.alifestd_to_working_format(draw_df)\n",
    "print(f\"{len(draw_df)=}\")\n",
    "draw_tree = apc.alife_dataframe_to_biopython_tree(\n",
    "    draw_df,\n",
    "    setattrs=[\"is_screened\", \"sister_screened\"],\n",
    "    setup_branch_lengths=True,\n",
    ")\n",
    "for node in draw_tree.find_clades():\n",
    "    node.color = [[\"black\", \"red\"], [\"blue\", \"green\"],][\n",
    "        node.is_screened\n",
    "    ][node.sister_screened]\n",
    "\n",
    "with tp.teed(\n",
    "    draw_biophylo,\n",
    "    draw_tree,\n",
    "    teeplot_outattrs={\n",
    "        \"color\": \"screened-sister\",\n",
    "        \"what\": slugify(str(mutation)),\n",
    "    },\n",
    "    teeplot_outdir=teeplot_subdir,\n",
    ") as ax:\n",
    "    ax.set_xlabel(\"Days\")\n",
    "    ax.yaxis.set_visible(False)\n",
    "    ax.spines[\"left\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "\n",
    "    legend_handles = [\n",
    "        mpl_patches.Patch(color=color, label=variant)\n",
    "        for variant, color in {\n",
    "            \"is_screened\": \"blue\",\n",
    "            \"sister_screened\": \"red\",\n",
    "        }.items()\n",
    "    ]\n",
    "    ax.legend(\n",
    "        handles=legend_handles,\n",
    "        loc=\"upper center\",\n",
    "        bbox_to_anchor=(0.5, 1.15),\n",
    "        ncol=2,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prep data\n",
    "data_df = phylo_df.copy()\n",
    "data_df[\"fisher trait-defined\"] = screen[\"fisher\"]\n",
    "data_df[\"naive trait-defined\"] = screen[\"naive\"]\n",
    "data_df[\"combined trait-defined\"] = screen[\"combined\"]\n",
    "screen_clade_size = (data_df[\"num_leaves\"] > 4) & (\n",
    "    data_df[\"num_leaves_sibling\"] > 4\n",
    ")\n",
    "data_df = data_df[screen_clade_size & ~data_df[\"is_root\"]].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot distributions\n",
    "for screen_trait_defined in (\n",
    "    # \"fisher trait-defined\",\n",
    "    # \"naive trait-defined\",\n",
    "    \"combined trait-defined\",\n",
    "):\n",
    "    print(data_df[screen_trait_defined].value_counts())\n",
    "    tp.tee(\n",
    "        mwu_kdeplot,\n",
    "        data_df,\n",
    "        x=\"num_leaves\",\n",
    "        hue=screen_trait_defined,\n",
    "        common_norm=False,\n",
    "        cut=True,\n",
    "        fill=True,\n",
    "        log_scale=(True, False),\n",
    "        teeplot_outattrs={\"mutation\": slugify(str(mutation))},\n",
    "        teeplot_outdir=teeplot_subdir,\n",
    "    )\n",
    "    tp.tee(\n",
    "        pairplot,\n",
    "        data_df,\n",
    "        hue=screen_trait_defined,\n",
    "        vars_to_plot=[\n",
    "            \"clade duration ratio\",\n",
    "            \"clade size ratio\",\n",
    "            \"clade growth ratio\",\n",
    "        ],\n",
    "        log_vars=[],\n",
    "        teeplot_outattrs={\"mutation\": slugify(str(mutation))},\n",
    "        teeplot_outdir=teeplot_subdir,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process Pathogen Phylogeny Statistics --- non-TFP\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mutation in it.islice(\n",
    "    ((site, from_, to) for site, from_, to in mutation_masks if site > 0),\n",
    "    10,\n",
    "):\n",
    "    display(HTML(f\"<h2>{mutation=}</h2>\"))\n",
    "\n",
    "    with hstrat_aux.log_context_duration(\n",
    "        \"screen_mutation_defined_nodes\",\n",
    "        logger=print,\n",
    "    ):\n",
    "        screen = screen_mutation_defined_nodes(\n",
    "            phylo_df.copy(),\n",
    "            has_mutation=mutation_masks[mutation],\n",
    "        )\n",
    "\n",
    "    print(\n",
    "        screen[\"fisher\"].sum(), screen[\"naive\"].sum(), screen[\"combined\"].sum()\n",
    "    )\n",
    "\n",
    "    data_df = phylo_df.copy()\n",
    "    data_df[\"fisher trait-defined\"] = screen[\"fisher\"]\n",
    "    data_df[\"naive trait-defined\"] = screen[\"naive\"]\n",
    "    data_df[\"combined trait-defined\"] = screen[\"combined\"]\n",
    "    screen_clade_size = (data_df[\"num_leaves\"] > min_leaves) & (\n",
    "        data_df[\"num_leaves_sibling\"] > min_leaves\n",
    "    )\n",
    "    data_df = data_df[screen_clade_size & ~data_df[\"is_root\"]].copy()\n",
    "\n",
    "    for screen_trait_defined in (\n",
    "        # \"fisher trait-defined\",\n",
    "        # \"naive trait-defined\",\n",
    "        \"combined trait-defined\",\n",
    "    ):\n",
    "        print(data_df[screen_trait_defined].value_counts())\n",
    "        tp.tee(\n",
    "            mwu_kdeplot,\n",
    "            data_df,\n",
    "            x=\"num_leaves\",\n",
    "            hue=screen_trait_defined,\n",
    "            common_norm=False,\n",
    "            cut=True,\n",
    "            fill=True,\n",
    "            log_scale=(True, False),\n",
    "            teeplot_outattrs={\"mutation\": slugify(str(mutation))},\n",
    "            teeplot_outdir=teeplot_subdir,\n",
    "        )\n",
    "        tp.tee(\n",
    "            pairplot,\n",
    "            data_df,\n",
    "            hue=screen_trait_defined,\n",
    "            vars_to_plot=[\n",
    "                \"clade duration ratio\",\n",
    "                \"clade size ratio\",\n",
    "                \"clade growth ratio\",\n",
    "            ],\n",
    "            log_vars=[],\n",
    "            teeplot_outattrs={\"mutation\": slugify(str(mutation))},\n",
    "            teeplot_outdir=teeplot_subdir,\n",
    "        )\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
